code/obj/item/assembly/misc_assemblies.dm

```diff
@@ -880,24 +880,39 @@ obj/item/assembly/radio_horn/receive_signal()
 //////////////////////////////////handmade shotgun shells//////////////////////////////////
 
 /obj/item/assembly/makeshiftshell
-
 	name = "filled pipe hulls"
 	desc = "Four open pipe shells, with propellant in them. You wonder what you could stuff into them."
 	icon_state = "Pipeshotrow"
+	var/thingsneeded = null //number of things still required to produce ammunition
+	var/crafttype = null
 
 	attackby(obj/item/W, mob/user)
-		if(istype(W, /obj/item/raw_material/shard))
-			var/obj/item/ammo/bullets/pipeshot/glass/shot = new /obj/item/ammo/bullets/pipeshot/glass/(get_turf(src))
-			qdel(W)
-			qdel(src)
-			user.put_in_hand_or_drop(shot)
-
-		if(istype(W, /obj/item/raw_material/scrap_metal))
-			var/obj/item/ammo/bullets/pipeshot/scrap/shot = new /obj/item/ammo/bullets/pipeshot/scrap/(get_turf(src))
-			qdel(W)
-			qdel(src)
-			user.put_in_hand_or_drop(shot)
+		if (!crafttype)
```

Comment:
My suggestion to make this attackby cleaner. Make the recipe handler a datum, with a child for each recipe:
1. When the first item is added, assign the datum
2. Let the datum handle inserting items into the shell (the logic of whether or not to accept the item, and whether or not it's done)
a. Datum should have a path to the craft material (shard / scrap_metal)
b. Datum should have a number needed of the craft material
c. Datum should have a path to the result ammo
d. Datum should have a proc to handle arbritrary attackby to check if it's the craft material
e. The proc in d should return whether or not the craft is done
4. When the datum returns that the craft is done, request the datum return the finished product and spawn it at the src location.

Reply:
good idea.